extension/src/documents.ts

```diff
@@ -5,9 +5,37 @@ import { AutoLispExt } from './extension';
 import { ProjectTreeProvider  } from "./project/projectTree";
 
 
+enum Origins {
+	OPENED = 'O',
+	WSPACE = 'W',
+	PROJECT = 'P'
+}
+
+
+interface DocumentSources {
+	native: vscode.TextDocument;
+	internal: ReadonlyDocument;
+	flags: Set<Origins>;
```

Comment:
It means that for one document, it can be opened, in workspace and in project files? maybe we need to add a comment here

Reply:
Well, think about life cycle and the goal to recycle everything. 
Opening a file adds the OPENED flag, which is the only source of truth for what is actively opened. That file still could have had WSPACE & PRJ flags.
If you delete a file using file explorer while it was opened in the editor, the WSPACE flag gets removed, but we still have the reference with the OPENED flag and we should because VSCode doesn't close the document in that scenario.
If you are in a workspace, you open a document from file explorer, you then decide you need to close it, but add it to a PRJ, then nothing is lost.
If you open a PRJ, use F12, close that PRJ open a different one and later come back to it or open a different PRJ that also references that file, then nothing is lost.

Reply:
Thanks for explaining it!